ArticleTemplates/assets/js/bootstraps/campaign.js

```diff
@@ -29,26 +51,44 @@ function initCampaign(campaign) {
 
     var form = campaign.querySelector('form');
     form.addEventListener('submit', function (e) {
+        displayWaiting(form);
         e.preventDefault();
         var data = Array.from(form.elements).reduce(function (o, e) {
             if (e.type === 'checkbox') {
                 if (e.checked) {
                     o[e.name] = o[e.name] ? o[e.name] + '\n' + e.value : e.value;
                 }
+            } else if (e.type === 'file') {
+                o[e.name] = readFile(e.files[0], campaign, form);
             } else if (e.value) {
                 o[e.name] = e.value;
             }
             
             return o;    
         }, {});
-        disableButton(form);
-        submit(data, campaign, form);
+
+        let promises = [];
+        let keys = [];
+        for (var key in data) {
```

Comment:
It feels odd to have this extra loop here. Shouldn't the loop that goes through each of the form element produce two arrays, one of field values, one of promises?